api/flask-app/app.py

```diff
@@ -11,64 +34,358 @@ def index():
-        return Response(status=200)
+        schema = UserSchema()
+        try:
+            user = schema.load(request.get_json())
+        except (ValidationError, TypeError):
+            return Response(status=400)
+        user_json = schema.dump(user)
+
+        with db.cursor() as cursor:
+            sql = "UPDATE User SET name=%(name)s, email=%(email)s, " \
+                  "bio=%(bio)s WHERE id=%(id)s"
+            args = user_json
+            args["id"] = id
+            cursor.execute(sql, args)
+
+        db.commit()
+        row = get_by_id(db.cursor(), "User", id)
+        return jsonify(row)
+
+    elif request.method == "DELETE":
+        delete_by_id(db.cursor(), "User", id)
+        db.commit()
+        return Response(status=204)
+
+    else:
+        # Unsupported HTTP method
+        return Response(status=405)
+
+
+@app.route("/users/<id>/followers", methods=["GET"])
+def followers(id):
+    db = get_db()
+    user = get_by_id(db.cursor(), "User", id)
+    if not user:
+        return Response(status=404)
+    with db.cursor() as cursor:
+        sql = "SELECT follower_id FROM Follow WHERE following_id=%s"
+        cursor.execute(sql, [id])
+        follower_ids = cursor.fetchall()
+
+        sql = "SELECT * FROM User WHERE id IN (%s)"
+        cursor.execute(
+            sql,
+            [', '.join(map(lambda x: str(x["follower_id"]), follower_ids))]
+        )
+        followers = cursor.fetchall()
+        return jsonify(followers)
+
+
+@app.route("/users/<id>/following", methods=["GET"])
+def following(id):
+    db = get_db()
+    user = get_by_id(db.cursor(), "User", id)
+    if not user:
+        return Response(status=404)
+    following_str = get_following_str(db.cursor(), id)
+
+    with db.cursor() as cursor:
+        sql = "SELECT * FROM User WHERE id IN (%s)"
+        cursor.execute(sql, [following_str])
+        following = cursor.fetchall()
+    return jsonify(following)
+
+
+@app.route("/users/<id>/feed", methods=["GET"])
+def feed(id):
+    db = get_db()
+    user = get_by_id(db.cursor(), "User", id)
+    if not user:
+        return Response(status=404)
+    following_str = get_following_str(db.cursor(), id)
+    with db.cursor() as cursor:
+        sql = "SELECT * FROM Message WHERE poster in (%s)"
+        cursor.execute(sql, [following_str])
+        feed = cursor.fetchall()
+    return jsonify(feed)
+
+
+@app.route("/messages", methods=["GET", "POST"])
+def messages():
+    db = get_db()
+
+    if request.method == "GET":
+        poster = request.args.get("poster")
+
+        with db.cursor() as cursor:
+            sql = "SELECT * FROM Message WHERE 1"
+            if poster:
+                sql += " AND poster=%(poster)s"
+            cursor.execute(sql, {"poster": poster})
+            messages = cursor.fetchall()
+
+        return jsonify(messages)
+
+    elif request.method == "POST":
+        schema = MessageSchema()
+        try:
+            message = schema.load(request.get_json())
```

Comment:
this schema.load function basically verifies that the input is formatted correctly right?

Reply:
Looks like this was the case. Adding `required=True` to each field stopped `TypeError`s being raised so they don't need to be caught.